presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java

```diff
@@ -123,14 +322,78 @@ private static OptionalLong reduce(OptionalLong first, OptionalLong second, Redu
                 case SUBTRACT:
                     return OptionalLong.of(first.getAsLong() - second.getAsLong());
+                case SELECT_MAX:
+                    return OptionalLong.of(max(first.getAsLong(), second.getAsLong()));
+                case SELECT_MIN:
+                    return OptionalLong.of(min(first.getAsLong(), second.getAsLong()));
+                case AVG:
+                    return OptionalLong.of((first.getAsLong() + second.getAsLong()) / 2);
+                default:
+                    throw new IllegalArgumentException("Unexpected operator: " + operator);
             }
         }
         return OptionalLong.empty();
     }
 
+    private static OptionalDouble reduce(OptionalDouble first, OptionalDouble second, ReduceOperator operator)
+    {
+        if (first.isPresent() && second.isPresent()) {
+            switch (operator) {
+                case ADD:
+                    return OptionalDouble.of(first.getAsDouble() + second.getAsDouble());
+                case SUBTRACT:
+                    return OptionalDouble.of(first.getAsDouble() - second.getAsDouble());
+                case SELECT_MAX:
+                    return OptionalDouble.of(max(first.getAsDouble(), second.getAsDouble()));
+                case SELECT_MIN:
+                    return OptionalDouble.of(min(first.getAsDouble(), second.getAsDouble()));
+                case AVG:
+                    return OptionalDouble.of((first.getAsDouble() + second.getAsDouble()) / 2);
+                default:
+                    throw new IllegalArgumentException("Unexpected operator: " + operator);
+            }
+        }
+        return OptionalDouble.empty();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static Optional<? extends Comparable<?>> reduce(
+            Optional<? extends Comparable<?>> firstOptional,
+            Optional<? extends Comparable<?>> secondOptional,
+            ReduceOperator operator)
+    {
+        if (firstOptional.isPresent() && secondOptional.isPresent()) {
+            Comparable<?> first = firstOptional.get();
+            Comparable<?> second = secondOptional.get();
+            checkArgument(first.getClass().equals(second.getClass()), "cannot compare comparable of different types: %s != %s", first.getClass(), second.getClass());
+            switch (operator) {
+                case SELECT_MAX:
+                    return Optional.of(max((Comparable) first, (Comparable) second));
+                case SELECT_MIN:
+                    return Optional.of(min((Comparable) first, (Comparable) second));
+                default:
+                    throw new IllegalArgumentException("Unexpected operator: " + operator);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static <T extends Comparable<T>> T max(T first, T second)
+    {
+        return first.compareTo(second) > 0 ? first : second;
```

Comment:
Might be better to use `>=` so that we return `first` on tie (even though they should be equal ... returning first seems more natural)